Support order-able and comparable variables in the simple function interface.

laithsakka opened the pull request:
Summary:
Before this diff, Generic<X> used to be translated to variables with the properties
only comparable and only order-able always being false.
this limitation is problematic for some functions like ArrayMax and Eq.

this diff allow expressions Orderable<Tx> and Comprable<Tx> in the simple function interface.

Differential Revision: D50568822

mbasmanova commented on velox/core/SimpleFunctionMetadata.h:
```diff
@@ -153,11 +154,22 @@ struct TypeAnalysisResults {
 
   // Set of generic variables used in the type.
-  std::set<std::string> variables;
+  std::map<std::string, exec::SignatureVariable> variablesInformation;
```
nit: std::unordered_map of F14FastMap?

laithsakka replied:
u mean to use folly::F14FastMap?

mbasmanova commented on velox/core/SimpleFunctionMetadata.h:
```diff
@@ -196,15 +208,21 @@ struct TypeAnalysis {
 
-template <typename T>
-struct TypeAnalysis<Generic<T>> {
+template <typename T, bool comparable, bool orderable>
```
perhaps, add a static assert that if orderable is true, then comparable is also true

laithsakka replied:
I already have it   in the class definition of Generic

template <typename T = AnyType, bool comparable = false, bool orderable = false>
struct Generic {
  Generic() = delete;
  static_assert(!(orderable && !comparable), "Orderable implies comparable.");
};

mbasmanova commented on velox/exec/tests/FunctionResolutionTest.cpp:
```diff
@@ -316,5 +319,119 @@ TEST_F(FunctionResolutionTest, resolveCustomTypeTimestampWithTimeZone) {
+// A function that takes TInput and return int, TInput determined at
+// registration.
+template <typename T>
+struct SignleInputFunc {
```
typo: Signle -> Single

mbasmanova commented on velox/exec/tests/FunctionResolutionTest.cpp:
```diff
@@ -316,5 +319,119 @@ TEST_F(FunctionResolutionTest, resolveCustomTypeTimestampWithTimeZone) {
   EXPECT_EQ(type->toString(), TIMESTAMP_WITH_TIME_ZONE()->toString());
 }
+
+// A function that takes TInput and return int, TInput determined at
```
return -> returns

mbasmanova reviewed (commented):
Looks good % a few nits. Please, update the documentation.

laithsakka commented:
@mbasmanova  @duanmeng thanks for the review addressed all comments
as for the documentation, i did not add any documentations for generic views, writers, or Generic type in the simple function
interface yet, hence there is nothing to update. 
Adding such documentation would be a separate task

mbasmanova commented on velox/exec/tests/FunctionResolutionTest.cpp:
```diff
@@ -316,5 +319,119 @@ TEST_F(FunctionResolutionTest, resolveCustomTypeTimestampWithTimeZone) {
+  testSingleInput(ROW({INTEGER(), MAP(INTEGER(), INTEGER())}));
+  testSingleInput(ARRAY(ARRAY(INTEGER())));
+
+  // Make sure we can assign different properties to the same variable.
```
can -> can't ?

mbasmanova reviewed (approved):
@laithsakka Thanks. Let's prioritize adding missing documentation for the SimpleFunction API.

mbasmanova commented:
CI is red.
